Fix 2087, 2088, and 2089 by properly setting parameters in rotation-commands and not using command feedback when relying on scratch-vm for motor timing

knandersen opened the pull request:
Resolves #2087 - Can't change motor direction while it's running
Resolves #2088 - Can't change motor power while it's running
Resolves #2089 - Default motor speed should be 50%?

Clarifies the renaming aspect mentioned in #2091

knandersen committed:
Resolves #2089.
Partly resolves #2087, #2088.

This was happening because scratch-vm is responsible for timed motor commands rather than using the Boost hubs commands to run motors for a specific amount of time.
This meant that when we simply sent a motorOn-command, the hub would immediately return feedback for the motor that the command had completed.
The fix for this was, for timed motor commands, to not receive feedback from the motor, and instead have scratch-vm modify the BoostMotor._status.

TODO: Fix for rotation-based commands.

Agreed. Changed to 50%.

- BoostMotorMaxPower changed to BoostMotorMaxPowerAdd and now describes an extra boost (no pun intended) to the motor. Documentation added that describes the feature.
- _colorBucket renamed to _colorSamples for clarity.
- oldColor is renamed to previousColor, and is now initialized in this._sensors.
- Modified documentation.

knandersen committed:
Resolves #2087 and #2088 for rotation-based commands.

There's quite a few interactions between degrees, their sign, and the currently set direction for the motor the degrees relate to. In this case, BoostMotor.turnOnDegrees() was being run with -degrees, and since that function does a Math.max between 0 and degrees, it resulted in 0 degrees.

Because of this, and for clarity, turnOnDegrees now only gets called with positive values. If running CCW, that should be specified in the direction-parameter.

knandersen committed:
- changed max power setting in motor functions to be calculated with MathUtil.clamp() instead of MathUtil.wrapClamp() to avoid values rolling over!
- added an else-case to both setMotorDirection() and setMotorPower() so that dynamic speed/direction changes also affect the motorOn()-blocks just like the time- and rotation-based ones.

evhan55 commented:
This tests ok for me after the latest commit!

- Can change direction on motor for all three motor on blocks (`seconds/rotations/on`)
- Can change power on motor for all three motor blocks (`seconds/rotations/on`)
- Motor default is at 50%

evhan55 commented on src/extensions/scratch3_boost/index.js:
```diff
@@ -590,15 +594,16 @@ class Boost {
          * @private
          */
-        this._colorBucket = [];
+        this._colorSamples = [];
```
Does this change belong in another PR, or should the issue this resolves be mentioned in the comment for this PR?

knandersen replied:
Yeah thanks for pointing that out @evhan55. It's from #2091!

evhan55 commented:
There is some code in this PR related to color sensing, should it be mentioned in the PR comment and reference an issue so I can test it?

evhan55 commented on src/extensions/scratch3_boost/index.js:
```diff
@@ -1926,10 +1934,10 @@ class Scratch3BoostBlocks {
         case BoostColorLabel.ANY:
             if (Object.keys(BoostColor).find(key => BoostColor[key])
                 .toLowerCase() !== this.getColor()) {
-                if (this.getColor() === this._peripheral.oldColor) {
```
Was this a bug before and should have been `this._peripheral._sensors.oldColor`? (Just trying to understand the change completely)

knandersen replied:
yes, you're right, it should be `this._peripheral._sensors.previousColor`(also updated the naming). It was addressed in commit e3cdbff

evhan55 commented:
The color name change should be okay, but I am not testing it since the color behavior is flaky at the moment in general for me, especially without LEGO bricks for testing.  I would say the color name change in this PR should be fine except for line #1929 which actually changes behavior a tiny bit.  If that is just a small bug fix, then it could land with this PR.  However, if it's important for me to test the change at line #1929, then I would maybe put all the color name change plus that bug fix in line #1929 in another PR later that addresses the color sensor changes if any will get made.  What do you think @knandersen ?

knandersen committed:
noticed several instances of getting especially power and direction properties from private variables instead of using the getter

knandersen committed:
It seems like there's a tradeoff between how we choose to set the max power of the motors. Previously, I had set the motors' max power (torque) to follow their target speed, meaning they accelerated smoothly, but also lost their torque. Then in the following commit I changed the max power to 100 which means maximum torque to achieve the target speed, which resulted in very abrupt accelerations and erratic motor movement when changing speeds:

https://github.com/LLK/scratch-vm/pull/2061/commits/873b56c9855a91b2d5e8797f9268406bcc8777c0

In the following commit I changed the functionality so that we add a fixed amount (10) more power than the target speed, e.g. for speed 50 it would provide power 60. This is fine for high speeds, but for low speeds it provides poor torque:

https://github.com/LLK/scratch-vm/pull/2100/commits/e3cdbffa2a96da31017e24484fbd75d21a19e2ad

I assume it would be possible to design some sort of calculation that enabled high torque for low speeds, and vice versa. I will discuss with the team.

knandersen committed:
changed getMotorPosition() to use the Boost.motor()-function instead of accessing the _motors-property directly

knandersen committed:
Merge branch 'develop' of https://github.com/LLK/scratch-vm into bugfix/2089,2088,2087

knandersen committed:
Added isBusy-flag in onMessage to make sure promises aren't resolved if the motor is still busy. Added check in motorOnForRotation() that ensure that any previous pendingPromiseFunction() is resolved before creating a new one, to avoid hanging blocks

knandersen committed:
Resolves #2087 and #2088. Because we weren't clearing a motor's _pendingPromiseFunction after executing it, it kept lingering, which made setMotorPower() and setMotorDirection() trigger a rotation-based command even if was responding to a timed or forever motorcommand. By clearing the property every time we fire the function, and by using pendingPromiseFunction as the conditional in setMotorPower() and setMotorDirection(), this should be taken care of.

evhan55 commented:
This feels mostly good to me now! 🎉 

- Can change motor direction for all motor on/seconds/rotations
- Can change motor power for all motor on/seconds/rotations
- Default was 50%
- Color bucket name change seems fine

Note:
- There is a tiny issue with rotations maybe not completing 100%, but I would rather find a repro for it and to resolve in a future PR, what do you think?

evhan55 reviewed (approved):
Looks mostly fine to me!  I think it might be good to merge at this step for testing at tomorrow's workshop.

knandersen commented:
Thanks @evhan55!
I had a really good chat with @ericrosenbaum earlier today about cleaning up handling what state the motor is in, since the code right now is error-prone and very hard to read.
I've tried to put our thinking in writing:

Since the motors can either really be off, on forever, on for an amount of time, or on for an amount of rotation, the idea would be to create a BoostMotorStatus-enum, with these values:

```
BoostMotorStatus {
  OFF: 0
  ON_FOREVER: 1
  ON_FOR_TIME: 2
  ON_FOR_ROTATION: 3
}
```

`BoostMotor.isOn()` would be replaced by `BoostMotor.getStatus()` which would get `BoostMotor._status` which can have a value from the `BoostMotorStatus`-enum.
A new function called `BoostMotor.setStatus(boostMotorStatus)` would be created, which would be responsible for not only giving the motor a status, but depending on the value supplied to it will clear the relevant motor-states, i.e.:

```
setStatus(state) {
  switch(state){
    case BoostMotorState.OFF:
      // Clear timed- and rotation-based motor-states
    case BoostMotorState.ON_FOREVER:
      // Clear rotation-based and timed motor-states
    case BoostMotorState.ON_FOR_TIME:
      // Clear rotation-based motor-states
    case BoostMotorState.ON_FOR_ROTATION:
      // Clear timed motor-states
  }
}
```

Thoughts? cc @cwillisf

evhan55 commented:
@knandersen 

Clean up sounds good as long as it keeps the functionality.  It would be easier for me to see if it makes the code more legible once this proposed enum is actually implemented.

One extra thought: my opinion would be to land the functionality as is, to keep a snapshot of "working code" which was tricky to get right. Then, make a new PR with the clean up steps, hopefully feeling safer to do that refactor knowing that you can always revert to working code.  Just my two cents!
